### PR TITLE
tsp, basic case for multipart HttpPart

### PR DESCRIPTION
**Pull request by weidongxu-microsoft**
link https://github.com/Azure/autorest.java/issues/2669

### CODE REVIEW

File: typespec-extension/src/common/client.ts [hunk 1]
```diff
@@ -50,9 +48,6 @@ export class ServiceVersion extends Metadata {
   }
 }
 
-export class ObjectScheme extends ObjectSchema {
+export interface CrossLanguageDefinition {
   crossLanguageDefinitionId?: string;
-  constructor(name: string, description: string, objectInitializer?: DeepPartial<ObjectSchema>) {
-    super(name, description, objectInitializer);
-  }
```
Comment (weidongxu-microsoft) [inline typespec-extension/src/common/client.ts:57]: This part of change is not related to multipart.

I was just surprised to see we had `ObjectScheme` and `ObjectSchema`.

I had plan to refactor the representation of `crossLanguageDefinitionId` in code model, but not having time though.

File: typespec-tests/tsp/multipart.tsp [hunk 2]
```diff
@@ -35,3 +35,25 @@ op upload(
...
+
+@route("/uploadHttpPart/images/{name}")
+@post
+op uploadHttpPart(
+  @path name: string,
+  @query compress?: boolean,
+  @header contentType: "multipart/form-data",
+  @multipartBody body: {
+    fileData1: HttpPart<Inherit2File>;
+    fileData2: HttpPart<File>;
```
Comment (weidongxu-microsoft) [inline typespec-tests/tsp/multipart.tsp:56]: 2 test cases.

1st for inherit and override of default property in `File`.
2nd for just using the `File` from http lib.

File: typespec-extension/src/external-schemas.ts [hunk 3]
```diff
@@ -108,30 +109,73 @@ export function createPollOperationDetailsSchema(schemas: Schemas, stringSchema:
...
+    // property.type is File, use name and properties from property.type for the File schema
+    /*
+    Current logic:
+    - Class name suffix "FileDetails"
+    - No class hierachy for File
+    - File has 3 properties: "content", "filename", "contentType" (Note that it is "contents" in TypeSpec)
+    - No adjustment on "content" property, it is always BinaryData and required
+    - Allow constant type for "filename" and "contentType" (to be discussed for other types e.g. enum)
+    - Allow required for "filename" and "contentType"
+     */
```
Comment (weidongxu-microsoft) [inline typespec-extension/src/external-schemas.ts:152]: I summarize current logic here. We would have more discussion in Scrum.

Comment (weidongxu-microsoft) [reply typespec-extension/src/external-schemas.ts:152]: Though I have certain logic (above) to handle some complex case, but the logic is not finalized, the goal of this PR is to allow service write basic `@multipartBody` and get good code generated, e.g.

```ts
  @multipartBody body: {
    file: HttpPart<File>
  }
```

and still generate reasonable code on complex cases.

Comment (weidongxu-microsoft) [reply typespec-extension/src/external-schemas.ts:159]: The rationale for
- Class name suffix "FileDetails"
is allow dev write `HttpPart<File>` and generate `FileDetails`

If we use the model name directly, user would need to do a `@clientName(File, "FileDetails", "java")`

File: typespec-extension/src/common/client.ts (hunk 1 above)
Comment (haolingdong-msft) [reply typespec-extension/src/common/client.ts:57]: Previously `ObjectScheme` is just used for adding `crossLanguageDefinitionId` ?

Comment (weidongxu-microsoft) [reply typespec-extension/src/common/client.ts:57]: That's my guess too.

For now, I changed to `(schema as CrossLanguageDefinition).crossLanguageDefinitionId = ...`

Future plan is to move this to `schema.language.default.crossLanguageDefinitionId` but that would involve Java changes.

File: typespec-extension/src/external-schemas.ts [hunk 4]
```diff
@@ -108,53 +115,164 @@ export function createPollOperationDetailsSchema(schemas: Schemas, stringSchema:
...
+      default: {
+        namespace: namespace,
+      },
+      java: {
+        namespace: getJavaNamespace(namespace),
+      },
+    },
+    serializationFormats: [KnownMediaType.Multipart],
+  });
+  fileDetailsSchema.serializationFormats;
```
Comment (XiaofeiCao) [inline typespec-extension/src/external-schemas.ts:141]: What does this line do?

File: typespec-extension/src/external-schemas.ts [hunk 5]
```diff
@@ -108,53 +115,164 @@ export function createPollOperationDetailsSchema(schemas: Schemas, stringSchema:
...
+  schemas.add(fileDetailsSchema);
+  fileDetailsMap.set(schemaName, fileDetailsSchema);
```
Comment (XiaofeiCao) [inline typespec-extension/src/external-schemas.ts:143]: Is this necessary? Seems all other places will set fileDetailsMap after calling this function.

Comment (weidongxu-microsoft) [reply typespec-extension/src/external-schemas.ts:143]: I think it is here
https://github.com/Azure/autorest.java/blob/14572ec8399685c072ac4366f889ede9ed0310f8/typespec-extension/src/external-schemas.ts#L224

If FileDetails of same name in the map, it won't get to this `createFileDetailsSchema` again.

File: typespec-extension/src/external-schemas.ts (hunk 4 above)
Comment (weidongxu-microsoft) [reply typespec-extension/src/external-schemas.ts:141]: Yeah, this is a mistake.

File: typespec-extension/src/external-schemas.ts (hunk 5 above)
Comment (XiaofeiCao) [reply typespec-extension/src/external-schemas.ts:142]: Yes, and we seem to have put into the map outside create function:
e.g. 
https://github.com/Azure/autorest.java/blob/14572ec8399685c072ac4366f889ede9ed0310f8/typespec-extension/src/external-schemas.ts#L227-L228

We could remove either one. Having them both seems excessive to me.

Comment (weidongxu-microsoft) [reply typespec-extension/src/external-schemas.ts:142]: Ah, I see. I thought `fileDetailsMap` is local var and won't be in the func. It is not.

File: typespec-extension/src/code-model-builder.ts [hunk 6]
```diff
@@ -2229,35 +2223,39 @@ export class CodeModelBuilder {
 
   private processMultipartFormDataFilePropertySchemaFromSdkType(
```
Comment (haolingdong-msft) [inline typespec-extension/src/code-model-builder.ts:2224]: So this function is not only for `FileProperty`? Do we want to rename the function to `processMultipartFormDataSchemaFromSdkType`?

Comment (weidongxu-microsoft) [reply typespec-extension/src/code-model-builder.ts:2224]: Currently it is only for File property
https://github.com/Azure/autorest.java/blob/6f322419cb50c6a342855ab70617bce96365e6cf/typespec-extension/src/code-model-builder.ts#L2104-L2106

Comment (haolingdong-msft) [reply typespec-extension/src/code-model-builder.ts:2224]: I'm a bit confused. Then why `getFileDetailsSchema` function can have 'model' type, I thought files are all bytes.

https://github.com/Azure/autorest.java/blob/6f322419cb50c6a342855ab70617bce96365e6cf/typespec-extension/src/external-schemas.ts#L209

Comment (weidongxu-microsoft) [reply typespec-extension/src/code-model-builder.ts:2224]: `HttpPart<File>`

that `File` is model

Comment (haolingdong-msft) [reply typespec-extension/src/code-model-builder.ts:2224]: I see, Thanks

File: typespec-extension/src/external-schemas.ts [hunk 7]
```diff
@@ -108,53 +115,161 @@ export function createPollOperationDetailsSchema(schemas: Schemas, stringSchema:
...
+      // crossLanguageDefinitionId
+      (fileDetailsSchema as CrossLanguageDefinition).crossLanguageDefinitionId =
+        property.type.crossLanguageDefinitionId;
+
+      let contentTypeProperty;
+      let filenameProperty;
+
+      // find "filename" and "contentType" property in current model and its base models
+      let type: SdkModelType | undefined = property.type;
+      while (type !== undefined) {
```
Comment (haolingdong-msft) [inline typespec-extension/src/external-schemas.ts:244]: I tend to let TCGC return the file related info, e.g. filename, headers(content-type or other headers), and we just use it. I'm fine if we handle it by ourselves currently, we can discuss later.

**Review by haolingdong-msft — APPROVED** [review]
Overall looks good to me.

File: typespec-extension/src/external-schemas.ts (hunk 7 above)
Comment (weidongxu-microsoft) [reply typespec-extension/src/external-schemas.ts:244]: 1. it could be in base class like `File`
2. it could be in multiple base class, if there is multiple `FileA extends FileB`, `FileB extends FileC`

Yeah, we may (or may not) be able to use the SdkModelProperty directly. E.g. I don't know what the SDK should generate, if typespec write
```
model FileA extends File {
  filename: "name1" | "name2"
}
```

Good news is that lint appears to error if adding any other property (other than this 3).
